Expensify/App
# Add logging to Network

@tgolen @marcaaron 

### Details

Added Logging to `Network.js` 
Added retry limit to failed `shouldRetry` requests 

### Fixed Issues

$ https://github.com/Expensify/App/issues/7666

### Tests


Find a way to make requests fail like switch network off in the middle of the request (it's hard to nail this)
Observe captured logs 
A request should not be retried more than 10 times

An easier (but lengthily) way to simulate this is 
1. Turn airplane mode on (or be offline)
2. Write a message 
3. Quit the app (terminate it) (you can resume connectivity after)
4. Wait for 2 hours so you're auth token expires
5. Launch the app
6. Observe logs for persisted request being retried (server logs)

#### Sample logs
#### 2 AddComment request being made and retried 
```
[info] Making API request - {"command":"Report_AddComment","type":"post","shouldUseSecure":false}
[info] Making API request - {"command":"Report_AddComment","type":"post","shouldUseSecure":false}
[info] Persisted request failed - {"retryCount":1,"command":"Report_AddComment","error":"Persisted request failed due to jsonCode: 407"}
[info] Persisted request failed - {"retryCount":1,"command":"Report_AddComment","error":"Persisted request failed due to jsonCode: 407"}
[info] Making API request - {"command":"Report_AddComment","type":"post","shouldUseSecure":false}
[info] Making API request - {"command":"Report_AddComment","type":"post","shouldUseSecure":false}
[info] Persisted request failed - {"retryCount":2,"command":"Report_AddComment","error":"Persisted request failed due to jsonCode: 407"}
[info] Persisted request failed - {"retryCount":2,"command":"Report_AddComment","error":"Persisted request failed due to jsonCode: 407"}
[info] Making API request - {"command":"Report_AddComment","type":"post","shouldUseSecure":false}
[info] Making API request - {"command":"Report_AddComment","type":"post","shouldUseSecure":false}
```

### QA Steps


Not sure whether QA can help here, maybe just check for regression related to networking
E.g. sending a message offline should still send the message after you resume connectivity 
It should be sent only one time

### Tested On

- ❌Desktop - untestable due to CORS issue

### Screenshots


#### Web

![image](https://user-images.githubusercontent.com/12156624/153443575-c9ad7db3-b2b9-4bdb-acc3-315ee2e8a857.png)

#### Mobile Web

![image](https://user-images.githubusercontent.com/12156624/153452611-d734c4fb-adcd-4176-8a42-ad253674007d.png)

#### Desktop


#### iOS

![image](https://user-images.githubusercontent.com/12156624/153452114-7d0ce37e-780c-4997-a1fd-50fb58faf888.png)
![image](https://user-images.githubusercontent.com/12156624/153452166-4c841eeb-03c7-43f0-8db6-bd0dd5897f7a.png)

#### Android

![image](https://user-images.githubusercontent.com/12156624/153458485-cf8fc300-3a55-4d8a-9b01-2e37ab9df2b4.png)
